### PR TITLE
instruction table: implement `trace_evaluation`

### PR DESCRIPTION
**Pull request by tcoratger**

### CODE REVIEW

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 1]
```diff
@@ -85,6 +110,68 @@ impl InstructionTable {
...
+    /// of all registers) to an array of columns (one element = all steps of one register).
+    /// It is then evaluated on the circle domain.
+    ///
+    /// # Returns
+    /// A tuple containing the evaluated trace and claim for STARK proof.
+    ///
+    /// # Errors
+    /// Returns [`TraceError::EmptyTrace`] if the table is empty.
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, Claim), TraceError> {
+        // Determine the number of rows in the table.
```
Comment (tcoratger) [inline crates/brainfuck_prover/src/components/instruction/table.rs:132]:
```suggestion
```

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 2]
```diff
@@ -125,6 +211,33 @@ impl From<(Vec<Registers>, &ProgramMemory)> for InstructionTable {
...
+    /// Index of the `ip` register column in the Instruction trace.
+    Ip,
+    /// Index of the `ci` register column in the Instruction trace.
+    Ci,
+    /// Index of the `ni` register column in the Instruction trace.
+    Ni,
+}
+
+impl InstructionColumn {
+    /// Returns the index of the column in the Instruction table
```
Comment (tcoratger) [inline crates/brainfuck_prover/src/components/instruction/table.rs:226]:
```suggestion
    /// Returns the index of the column in the Instruction trace
```

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 3]
```diff
@@ -125,6 +211,33 @@ impl From<(Vec<Registers>, &ProgramMemory)> for InstructionTable {
...
+    /// Returns the index of the column in the Instruction trace
+    pub const fn index(self) -> usize {
+        match self {
+            Self::Ip => 0,
+            Self::Ci => 1,
+            Self::Ni => 2,
+        }
+    }
+
+    /// Returns the total number of columns in the Instruction table
```
Comment (tcoratger) [inline crates/brainfuck_prover/src/components/instruction/table.rs:235]:
```suggestion
    /// Returns the total number of columns in the Instruction trace
```

File: crates/brainfuck_prover/src/components/instruction/table.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/instruction/table.rs]: done

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 4]
```diff
@@ -125,6 +212,33 @@ impl From<(Vec<Registers>, &ProgramMemory)> for InstructionTable {
...
+    /// Index of the `ip` register column in the Instruction trace.
+    Ip,
+    /// Index of the `ci` register column in the Instruction trace.
+    Ci,
+    /// Index of the `ni` register column in the Instruction trace.
+    Ni,
+}
+
+impl InstructionColumn {
+    /// Returns the index of the column in the Instruction table
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/instruction/table.rs]: done

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 5]
```diff
@@ -125,6 +212,33 @@ impl From<(Vec<Registers>, &ProgramMemory)> for InstructionTable {
...
+    pub const fn index(self) -> usize {
+        match self {
+            Self::Ip => 0,
+            Self::Ci => 1,
+            Self::Ni => 2,
+        }
+    }
+
+    /// Returns the total number of columns in the Instruction table
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/instruction/table.rs]: done

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 6]
```diff
@@ -85,6 +110,68 @@ impl InstructionTable {
...
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, Claim), TraceError> {
+        // Determine the number of rows in the table.
+        let n_rows = self.table.len() as u32;
+        // If the table is empty, there is no data to evaluate, so return an error.
+        if n_rows == 0 {
+            return Err(TraceError::EmptyTrace);
+        }
+
+        // Compute `log_n_rows`, the base-2 logarithm of the number of rows.
+        // This determines the smallest power of two greater than or equal to `n_rows`.
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/instruction/table.rs]: added to the doc

File: crates/brainfuck_prover/src/components/instruction/table.rs [hunk 7]
```diff
@@ -365,4 +479,130 @@ mod tests {
...
+
+        assert_eq!(claim.log_size, LOG_N_LANES, "Log size should include SIMD lanes.");
+        assert_eq!(
+            trace.len(),
+            InstructionColumn::count(),
+            "Trace should contain one column per register."
+        );
+
+        // Check that each column contains the correct values
+        assert_eq!(trace[InstructionColumn::Ip.index()].to_cpu().values[0], BaseField::from(1));
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/instruction/table.rs]: done, as this test is simple, I've directly hardcoded 16, the expected number of elements, let me know if this is ok for you

**Review by zmalatrax — APPROVED** [review]

**Pull request merged**